### PR TITLE
Initial Commit: Exercise 03

### PR DESCRIPTION
**Pull request by RubelOS**
The Result to Subtask 3 is: 1306

### CODE REVIEW

**Comment by GeFritz** [comment]
Wieso hast du ein pull request eröffnet obwohl du nicht fertig bist?

**Comment by RubelOS** [comment]
True, hätte nur den Branch eröffnen können. War etwas voreiligAm 22.10.2023 01:47 schrieb Gerrit ***@***.***>:
Wieso hast du ein pull request eröffnet obwohl du nicht fertig bist?

File: Exercises/JuliusDralle/exercises_03.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                    #print("adjacence given")
+        if adjacence == True:
+            if number_string == number_string_sorted:
+                isValid = True
+        if isValid:
+            validNumbers.append(number_string)
+    return(len(validNumbers))
+
+# print(numberCriteriaCheckerCounter(134564,585159))
+# The Result to this is: 19
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:179]: The answer to your snippet is 1992.

File: Exercises/JuliusDralle/exercises_03.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    """
+    This function takes two arguments: lower bound and higher bound, both as int
+    This function first tests if a number contains a adjacent pair of digits.
+    After that is tests if the number is sorted already
+    If both tests come out positive, the amount of numbers in the bounds is being represented
+    """
+    bound_lower = int(bound_lower)
+    bound_higher = int(bound_higher)
+    validNumbers = []
+    for iterator_numbers in range(bound_lower, bound_higher+1):
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:161]: The task states including lower bound and excluding upper bound.

File: Exercises/JuliusDralle/exercises_03.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        number_string = str(iterator_numbers)
+        number_string_sorted = "".join(sorted(number_string))
+        adjacence = False
+        isValid = False
+        for iterator_charsInString in range(0, len(number_string)):
+            if iterator_charsInString > 0:
+                if number_string[iterator_charsInString - 1] == number_string[iterator_charsInString]:
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:168]: There is a mistake in your code as you also count numbers in rows of three and higher. The task explicitly states an exact amount of two equal numbers. 
If we add `and number_string.count(number_string[iterator_charsInString])==2` in line 168, the answer is `1306 `which seems to be correct.

File: Exercises/JuliusDralle/exercises_03.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    "08" : "Die Acht (8): Die Acht symbolisiert Fülle und Überfluss. Wie die Unendlichkeitsschleife erinnert sie uns daran, dass wir nie aufhören sollten, nach Erfolg und Wohlstand zu streben.",
+    "09" : "Die Neun (9): Die Neun steht für Vollendung und Erfüllung. Wie die Monate in einem Jahr erinnert sie uns daran, dass wir unsere Ziele erreichen können, wenn wir hart arbeiten und Ausdauer zeigen.",
+    "10" : "Die Zehn (10): Die Zehn symbolisiert Vollständigkeit und Erfolg. Wie die Finger an unseren Händen erinnert sie uns daran, dass wir alle Ressourcen zur Verfügung haben, um unsere Träume zu verwirklichen.",
+    "11" : "Der Bube (Jack): Der Bube verkörpert Jugend und Abenteuerlust. Wie ein mutiger Held auf einer Reise erinnert er uns daran, dass wir Risiken eingehen und die Welt entdecken sollten.",
+    "12" : "Die Dame (Queen): Die Dame steht für Anmut und Weisheit. Wie eine Königin auf ihrem Thron erinnert sie uns daran, dass wir unser Leben mit Würde und Klugheit führen sollten.",
+    "12" : "Der König (King): Der König symbolisiert Macht und Autorität. Wie ein weiser Herrscher erinnert er uns daran, Verantwortung zu übernehmen und unser Leben zu regieren."
+}
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:46]: That is alot of work. But it looks nice, well done!

File: Exercises/JuliusDralle/exercises_03.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+# The deck of cards should behave like a sequence.
+# When initialized the cards should all be in a well-defined order (2-Ace of each suite, suites in the order above
+# I should be able to index into the deck to retrieve one card.
+# I should be able to iterate over all cards in the deck.
+# Printing a cards string representation should give me a nice, 
+# readable description of that card.
+
+cardtype_dict = {
+    "01" : "Ace",
+    "02" : "2",
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:12]: In the task 2-Ace means in range 2 to ace, not 2-ace of diamonds.

File: Exercises/JuliusDralle/exercises_03.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        self.card_type = card_type_id
+
+    def __str__(self):
+        return(description_dict[self.card_type])
+    
+    def print_value(self):
+        print(f"This card is a \t{cardtype_dict[self.card_type]} \tin this suite: {suits_dict[self.card_suite]}")
+
+    def get_value(self):
+        return(cardtype_dict[self.card_type], suits_dict[self.card_suite])
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:92]: Nice solution to implement new card class. Using dict ids is also simple and clear. What if index is out of range?

File: Exercises/JuliusDralle/exercises_03.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+# PART 2:
+# Create a second class that represents a deck of cards usable for Skat -- it should only contain cards from 7 upwards.
+# It should offer all the same functionality of the first class.
+
+class DeckOfSkatCards:
+    """
+    The DeckOfSkatCards class is a Deck of cards for a Skat game. It only contains cards from the 7 up beginning
+    """
+    wanted_cards = ["07","08","09","10","11","12"]
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:109]: The skat deck is missing the ace card for each suit. It should be 32 cards in total.

**Review by GeFritz — CHANGES_REQUESTED** [review]
The code `2-Ace of diamonds ` seems to be a misunderstanding. Else nice and readable code. Answer to question three is wrong, lacks just one simple logic fix. Consider using assert functions to test your code.

File: Exercises/JuliusDralle/exercises_03.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    "04" : "4",
+    "05" : "5",
+    "06" : "6",
+    "07" : "7",
+    "08" : "8",
+    "09" : "9",
+    "10" : "10",
+    "11" : "Jack",
+    "12" : "Queen",
+    "12" : "King"
```
Comment (GeFritz) [inline Exercises/JuliusDralle/exercises_03.py:23]: There seems to be a key error here.

File: Exercises/JuliusDralle/exercises_03.py (hunk 1 above)
Comment (RubelOS) [reply Exercises/JuliusDralle/exercises_03.py:179]: Yes sorry, i did miss the last two chars. Thanks!

File: Exercises/JuliusDralle/exercises_03.py (hunk 2 above)
Comment (RubelOS) [reply Exercises/JuliusDralle/exercises_03.py:161]: I did not even notice. Thank you!

File: Exercises/JuliusDralle/exercises_03.py (hunk 4 above)
Comment (RubelOS) [reply Exercises/JuliusDralle/exercises_03.py:46]: I gave Copyright to ChatGPT for creating those values. Thank you!

File: Exercises/JuliusDralle/exercises_03.py (hunk 5 above)
Comment (RubelOS) [reply Exercises/JuliusDralle/exercises_03.py:12]: What exactly do you mean with this comment?

File: Exercises/JuliusDralle/exercises_03.py (hunk 6 above)
Comment (RubelOS) [reply Exercises/JuliusDralle/exercises_03.py:92]: It cannot be out of range. The Deck iterates through two dictionaries which are just beind send over to the card objects

**Review by GeFritz — CHANGES_REQUESTED** [review]
Just change line 26 then done. Sorry if i am being strict but am just doing my job. Very nice idea with chat gpt tho.

**Review by GeFritz — APPROVED** [review]

**Pull request merged**